src/main/java/gillycarpetaddons/mixins/SpectatorPortalFix.java

```diff
@@ -0,0 +1,55 @@
+package gillycarpetaddons.mixins;
+
+import com.mojang.authlib.GameProfile;
+import gillycarpetaddons.GillyCarpetAddonsSettings;
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.entity.MovementType;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.World;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+
+@Mixin(ServerPlayerEntity.class)
+public abstract class SpectatorPortalFix extends PlayerEntity {
+    public SpectatorPortalFix(World world, BlockPos pos, float yaw, GameProfile gameProfile) {
+        super(world, pos, yaw, gameProfile);
+    }
+
+    @Shadow public abstract boolean isSpectator();
+
+    /*
+        Calls original move() function then performs additional action
+        this.getBlockPos() returns coordinates of the lowest block that spectator's bounding box would be touching
+        if it was in survival game-mode at the same coordinates with default pose.
+     */
+    @Override
+    public void move(MovementType type, Vec3d movement) {
+        super.move(type,movement);
+        if(GillyCarpetAddonsSettings.spectatorPortalFix && this.isSpectator()){
```

Comment:
This is more of a personal preference (but still favoured by a lot of developers) but the more indented if statements there are, the harder it gets reading the code. I'd consider inverting this if statement so you return immediately if the player isn't a spectator or the rule is disabled.